File: go/17-produccion/mini-incident-sim.go
// Mini-proyecto: El simulador de incidentes de producción (Go).
//
// Uso:
//
//	go run mini-incident-sim.go
//	go run mini-incident-sim.go -fallo timeout
//	go run mini-incident-sim.go -fallo todos

// Cómo ejecutar: make go FILE=go/17-produccion/mini-incident-sim.go

package main

import (
	"flag"
	"fmt"
	"math"
	"math/rand"
	"strings"
)

// ── tipos ─────────────────────────────────────────────────────────────────────

type Resultado struct {
	exito     bool
	intentos  int
	durMs     float64
	error     string
	estrategia string
	detalles  []string
}

// ── circuit breaker ────────────────────────────────────────────────────────────

type CircuitBreaker struct {
	nombre       string
	umbralFallos int
	fallos       int
	estado       string // closed | open | half-open
}

func newCB(nombre string, umbral int) *CircuitBreaker {
	return &CircuitBreaker{nombre: nombre, umbralFallos: umbral, estado: "closed"}
}

func (cb *CircuitBreaker) puedesPasar() bool {
	return cb.estado != "open"
}

func (cb *CircuitBreaker) registrarExito() {
	cb.fallos = 0
	cb.estado = "closed"
}

func (cb *CircuitBreaker) registrarFallo() {
	cb.fallos++
	if cb.fallos >= cb.umbralFallos {
		cb.estado = "open"
	}
}

// ── estrategias ───────────────────────────────────────────────────────────────

func jitter(baseMs float64, intento int, rng *rand.Rand) float64 {
	espera := baseMs * math.Pow(2, float64(intento))
	j := espera * 0.25 * (rng.Float64()*2 - 1)
	if espera+j < 100 {
		return 100
	}
	return espera + j
}

func recuperarTimeout(rng *rand.Rand) Resultado {
	var detalles []string
	for i := 0; i < 3; i++ {
		espera := jitter(500, i, rng)
		detalles = append(detalles, fmt.Sprintf("  Intento %d: espera %.0fms antes de reintentar", i+1, espera))
		if rng.Float64() > 0.4 {
			detalles = append(detalles, fmt.Sprintf("  ✓ LLM completada en intento %d", i+1))
			return Resultado{exito: true, intentos: i + 1, durMs: 1200 + float64(i)*800,
				estrategia: "retry_con_jitter", detalles: detalles}
		}
	}
	detalles = append(detalles, "  ✗ Todos los reintentos agotados")
	return Resultado{exito: false, intentos: 3, durMs: 5000,
		error: "LLM timeout tras 3 reintentos", estrategia: "retry_con_jitter", detalles: detalles}
}

func recuperarOutputMalformado(rng *rand.Rand) Resultado {
	detalles := []string{
		"  Output: '{\"hallazgos\": [broken json...'",
		"  Detección: JSONDecodeError",
		"  Estrategia: feedback al modelo con el error exacto",
		"  Prompt recovery: 'Tu respuesta no era JSON válido. Responde SOLO con JSON...'",
	}
	if rng.Float64() > 0.2 {
		detalles = append(detalles, "  ✓ Segundo intento produjo JSON válido")
		return Resultado{exito: true, intentos: 2, durMs: 1400,
			estrategia: "feedback_al_modelo", detalles: detalles}
	}
	detalles = append(detalles, "  ✗ Segundo intento también malformado")
	return Resultado{exito: false, intentos: 2, durMs: 1800,
		error: "Output malformado en 2 intentos", estrategia: "feedback_al_modelo", detalles: detalles}
}

func recuperarContextOverflow(tokensActuales, ventana int) Resultado {
	usoPct := float64(tokensActuales) / float64(ventana) * 100
	detalles := []string{
		fmt.Sprintf("  Contexto: %d tokens (%.1f%% de %d)", tokensActuales, usoPct, ventana),
	}
	if usoPct > 75 {
		objetivo := int(float64(ventana) * 0.6)
		liberados := tokensActuales - objetivo
		detalles = append(detalles,
			"  Umbral 75% superado — clearing de tool results",
			fmt.Sprintf("  Tokens liberados: ~%d", liberados),
			fmt.Sprintf("  Contexto resultante: %d (%.1f%%)", objetivo, float64(objetivo)/float64(ventana)*100),
		)
		return Resultado{exito: true, intentos: 1, durMs: 50,
			estrategia: "compresion_proactiva_75pct", detalles: detalles}
	}
	detalles = append(detalles, "  Dentro de límites — no requiere compresión.")
	return Resultado{exito: true, intentos: 1, durMs: 10, estrategia: "ninguna", detalles: detalles}
}

func recuperarToolFallo(cb *CircuitBreaker, rng *rand.Rand) Resultado {
	detalles := []string{
		fmt.Sprintf("  Circuit breaker '%s': estado=%s, fallos=%d", cb.nombre, cb.estado, cb.fallos),
	}
	if !cb.puedesPasar() {
		detalles = append(detalles,
			"  ✗ Circuit breaker ABIERTO — herramienta no disponible",
			"  Fallback: usando caché o resultado por defecto",
		)
		return Resultado{exito: false, intentos: 0, durMs: 5,
			error: fmt.Sprintf("Circuit breaker abierto para '%s'", cb.nombre),
			estrategia: "circuit_breaker_fallback", detalles: detalles}
	}
	detalles = append(detalles, "  Circuit breaker cerrado — intentando llamada")
	if rng.Float64() > 0.5 {
		cb.registrarExito()
		detalles = append(detalles, fmt.Sprintf("  ✓ Herramienta '%s' respondió", cb.nombre))
		return Resultado{exito: true, intentos: 1, durMs: 200,
			estrategia: "circuit_breaker_normal", detalles: detalles}
	}
	cb.registrarFallo()
	detalles = append(detalles, fmt.Sprintf("  ✗ Fallo — acumulados: %d/%d", cb.fallos, cb.umbralFallos))
	if cb.estado == "open" {
		detalles = append(detalles, "  Circuit breaker ahora ABIERTO")
	}
	return Resultado{exito: false, intentos: 1, durMs: 5000,
		error: "Tool timeout", estrategia: "circuit_breaker_normal", detalles: detalles}
}

func recuperarBudgetExcedido(costoActual, budget float64) Resultado {
	exceso := costoActual - budget
	detalles := []string{
		fmt.Sprintf("  Costo acumulado: $%.4f", costoActual),
		fmt.Sprintf("  Budget de tarea: $%.4f", budget),
		fmt.Sprintf("  Exceso: $%.4f (%.0f%% sobre budget)", exceso, exceso/budget*100),
	}
	if costoActual > budget {
		detalles = append(detalles,
			"  Estrategia: degradación a modelo económico para pasos restantes",
			"  Haiku ($0.80/Mtok) reemplaza Sonnet ($3.00/Mtok) — 3.75× más barato",
		)
		return Resultado{exito: true, intentos: 1, durMs: 0,
			estrategia: "model_downgrade_budget", detalles: detalles}
	}
	detalles = append(detalles, "  Budget no excedido.")
	return Resultado{exito: true, intentos: 1, durMs: 0, estrategia: "ninguna", detalles: detalles}
}

// ── presentación ──────────────────────────────────────────────────────────────

func imprimirResultado(tipo string, r Resultado) {
	estado := "✓ RECUPERADO"
	if !r.exito {
		estado = "✗ FALLIDO"
	}
	fmt.Printf("\n  %s\n", strings.Repeat("─", 56))
	fmt.Printf("  Fallo: %s\n", strings.ToUpper(tipo))
	fmt.Printf("  Estado: %s  |  Estrategia: %s\n", estado, r.estrategia)
	fmt.Printf("  Intentos: %d  |  Duración: %.0fms\n", r.intentos, r.durMs)
	if r.error != "" {
		fmt.Printf("  Error final: %s\n", r.error)
	}
	fmt.Println("\n  Traza de recuperación:")
	for _, d := range r.detalles {
		fmt.Printf("  %s\n", d)
	}
}

func imprimirResumen(fallos []string, resultados map[string]Resultado) {
	recuperados := 0
	for _, r := range resultados {
		if r.exito {
			recuperados++
		}
	}
	fmt.Printf("\n%s\n", strings.Repeat("=", 60))
	fmt.Println("  RESUMEN — Simulador de Incidentes de Producción")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("\n  %d/%d fallos recuperados automáticamente\n", recuperados, len(resultados))
	fmt.Printf("\n  %-22s %-16s %s\n", "Fallo", "Estado", "Estrategia")
	fmt.Printf("  %s\n", strings.Repeat("─", 56))
	for _, f := range fallos {
		r := resultados[f]
		estado := "RECUPERADO"
		if !r.exito {
			estado = "FALLIDO"
		}
		fmt.Printf("  %-22s %-16s %s\n", f, estado, r.estrategia)
	}
	fmt.Println("\n  Lecciones clave:")
	fmt.Println("  • Timeout: retry con jitter previene thundering herd")
	fmt.Println("  • Output malformado: feedback exacto al modelo")
	fmt.Println("  • Context overflow: compresión al 75% de uso, no al 100%")
	fmt.Println("  • Tool fallo: circuit breaker evita cascada de timeouts")
	fmt.Println("  • Budget excedido: degradar modelo antes de abortar")
	fmt.Println(strings.Repeat("=", 60))
}

// ── main ──────────────────────────────────────────────────────────────────────

func main() {
	fallo := flag.String("fallo", "todos", "Tipo: timeout|output_malformado|context_overflow|tool_fallo|budget_excedido|todos")
	flag.Parse()

	rng := rand.New(rand.NewSource(42))
	cb := newCB("search_docs", 3)
	cb.fallos = 3
	cb.estado = "open"

	fmt.Printf("\n%s\n", strings.Repeat("=", 60))
	fmt.Println("  SIMULADOR DE INCIDENTES DE PRODUCCIÓN")
	fmt.Printf("  Fallo: %s\n", *fallo)
	fmt.Println(strings.Repeat("=", 60))

	todos := []string{"timeout", "output_malformado", "context_overflow", "tool_fallo", "budget_excedido"}
	var fallos []string
	if *fallo == "todos" {
		fallos = todos
	} else {
		fallos = []string{*fallo}
	}

	resultados := make(map[string]Resultado)
	for _, f := range fallos {
		var r Resultado
		switch f {
		case "timeout":
			r = recuperarTimeout(rng)
		case "output_malformado":
			r = recuperarOutputMalformado(rng)
		case "context_overflow":
			r = recuperarContextOverflow(7200, 8192)
		case "tool_fallo":
			r = recuperarToolFallo(cb, rng)
		case "budget_excedido":
			r = recuperarBudgetExcedido(0.082, 0.06)
		default:
			continue
		}
		imprimirResultado(f, r)
		resultados[f] = r
	}

	if len(resultados) > 1 {
		imprimirResumen(fallos, resultados)
	}
}

File: typescript/17-produccion/mini-incident-sim.ts
/**
 * Mini-proyecto: El simulador de incidentes de producción (TypeScript).
 *
 * Uso:
 *   npx ts-node mini-incident-sim.ts
 *   npx ts-node mini-incident-sim.ts --fallo timeout
 *   npx ts-node mini-incident-sim.ts --fallo todos
 */

interface Resultado {
  exito: boolean;
  intentos: number;
  duracionMs: number;
  error: string;
  estrategia: string;
  detalles: string[];
}

// ── circuit breaker ────────────────────────────────────────────────────────────

// Cómo ejecutar: make ts SCRIPT=typescript/17-produccion/mini-incident-sim.ts


class CircuitBreaker {
  private fallos = 0;
  private estado: "closed" | "open" | "half-open" = "closed";
  private ultimoFallo = 0;

  constructor(readonly nombre: string, readonly umbralFallos = 3, readonly timeoutS = 30) {}

  puedesPasar(): boolean {
    if (this.estado === "closed") return true;
    if (this.estado === "open") {
      if (Date.now() / 1000 - this.ultimoFallo > this.timeoutS) { this.estado = "half-open"; return true; }
      return false;
    }
    return true;
  }

  registrarExito(): void { this.fallos = 0; this.estado = "closed"; }

  registrarFallo(): void {
    this.fallos++;
    this.ultimoFallo = Date.now() / 1000;
    if (this.fallos >= this.umbralFallos) this.estado = "open";
  }

  get estadoActual(): string { return this.estado; }
  get fallosAcumulados(): number { return this.fallos; }
  set fallosExterno(n: number) { this.fallos = n; if (n >= this.umbralFallos) this.estado = "open"; }
}

// ── generador aleatorio simple ────────────────────────────────────────────────

let seed = 42;
function rng(): number { seed = (seed * 1664525 + 1013904223) >>> 0; return seed / 0xffffffff; }

// ── estrategias ───────────────────────────────────────────────────────────────

function jitter(baseMs: number, intento: number): number {
  const espera = baseMs * Math.pow(2, intento);
  return Math.max(100, espera + espera * 0.25 * (rng() * 2 - 1));
}

function recuperarTimeout(maxIntentos = 3): Resultado {
  const detalles: string[] = [];
  for (let i = 0; i < maxIntentos; i++) {
    const espera = jitter(500, i);
    detalles.push(`  Intento ${i + 1}: espera ${espera.toFixed(0)}ms antes de reintentar`);
    if (rng() > 0.4) {
      detalles.push(`  ✓ Llamada LLM completada en intento ${i + 1}`);
      return { exito: true, intentos: i + 1, duracionMs: 1200 + i * 800, error: "", estrategia: "retry_con_jitter", detalles };
    }
  }
  detalles.push("  ✗ Todos los reintentos agotados");
  return { exito: false, intentos: maxIntentos, duracionMs: 5000, error: "LLM timeout tras 3 reintentos", estrategia: "retry_con_jitter", detalles };
}

function recuperarOutputMalformado(): Resultado {
  const detalles = [
    "  Output recibido: '{\"hallazgos\": [broken json...'",
    "  Detección: JSONParseError",
    "  Estrategia: feedback al modelo con el error exacto",
    "  Prompt: 'Tu respuesta anterior no era JSON válido. Error: JSONParseError. Responde SOLO con JSON...'",
  ];
  if (rng() > 0.2) {
    detalles.push("  ✓ Segundo intento produjo JSON válido");
    return { exito: true, intentos: 2, duracionMs: 1400, error: "", estrategia: "feedback_al_modelo", detalles };
  }
  detalles.push("  ✗ Segundo intento también malformado");
  return { exito: false, intentos: 2, duracionMs: 1800, error: "Output malformado en 2 intentos", estrategia: "feedback_al_modelo", detalles };
}

function recuperarContextOverflow(tokensActuales: number, ventana: number): Resultado {
  const usoPct = tokensActuales / ventana * 100;
  const detalles = [`  Contexto: ${tokensActuales.toLocaleString()} tokens (${usoPct.toFixed(1)}% de ${ventana.toLocaleString()})`];
  if (usoPct > 75) {
    const objetivo = Math.floor(ventana * 0.6);
    const liberados = tokensActuales - objetivo;
    detalles.push(`  Umbral 75% superado — clearing de tool results`);
    detalles.push(`  Tokens liberados: ~${liberados.toLocaleString()}`);
    detalles.push(`  Contexto resultante: ${objetivo.toLocaleString()} (${(objetivo / ventana * 100).toFixed(1)}%)`);
    return { exito: true, intentos: 1, duracionMs: 50, error: "", estrategia: "compresion_proactiva_75pct", detalles };
  }
  detalles.push("  Dentro de límites — no requiere compresión.");
  return { exito: true, intentos: 1, duracionMs: 10, error: "", estrategia: "ninguna", detalles };
}

function recuperarToolFallo(cb: CircuitBreaker): Resultado {
  const detalles = [`  Circuit breaker '${cb.nombre}': estado=${cb.estadoActual}, fallos=${cb.fallosAcumulados}`];
  if (!cb.puedesPasar()) {
    detalles.push("  ✗ Circuit breaker ABIERTO — herramienta no disponible");
    detalles.push("  Fallback: usando caché o resultado por defecto");
    return { exito: false, intentos: 0, duracionMs: 5, error: `Circuit breaker abierto para '${cb.nombre}'`, estrategia: "circuit_breaker_fallback", detalles };
  }
  detalles.push("  Circuit breaker cerrado — intentando llamada");
  if (rng() > 0.5) {
    cb.registrarExito();
    detalles.push(`  ✓ Herramienta '${cb.nombre}' respondió correctamente`);
    return { exito: true, intentos: 1, duracionMs: 200, error: "", estrategia: "circuit_breaker_normal", detalles };
  }
  cb.registrarFallo();
  detalles.push(`  ✗ Fallo — acumulados: ${cb.fallosAcumulados}/${cb.umbralFallos}`);
  if (cb.estadoActual === "open") detalles.push(`  Circuit breaker ahora ABIERTO`);
  return { exito: false, intentos: 1, duracionMs: 5000, error: "Tool timeout", estrategia: "circuit_breaker_normal", detalles };
}

function recuperarBudgetExcedido(costoActual: number, budget: number): Resultado {
  const exceso = costoActual - budget;
  const detalles = [
    `  Costo acumulado: $${costoActual.toFixed(4)}`,
    `  Budget de tarea: $${budget.toFixed(4)}`,
    `  Exceso: $${exceso.toFixed(4)} (${(exceso / budget * 100).toFixed(0)}% sobre budget)`,
  ];
  if (costoActual > budget) {
    detalles.push("  Estrategia: degradación a modelo económico para pasos restantes");
    detalles.push("  Haiku ($0.80/Mtok) reemplaza a Sonnet ($3.00/Mtok) — 3.75× más barato");
    return { exito: true, intentos: 1, duracionMs: 0, error: "", estrategia: "model_downgrade_budget", detalles };
  }
  detalles.push("  Budget no excedido.");
  return { exito: true, intentos: 1, duracionMs: 0, error: "", estrategia: "ninguna", detalles };
}

// ── presentación ──────────────────────────────────────────────────────────────

function imprimirResultado(tipo: string, r: Resultado): void {
  console.log(`\n  ${"─".repeat(56)}`);
  console.log(`  Fallo: ${tipo.toUpperCase()}`);
  console.log(`  Estado: ${r.exito ? "✓ RECUPERADO" : "✗ FALLIDO"}  |  Estrategia: ${r.estrategia}`);
  console.log(`  Intentos: ${r.intentos}  |  Duración: ${r.duracionMs.toFixed(0)}ms`);
  if (r.error) console.log(`  Error final: ${r.error}`);
  console.log(`\n  Traza de recuperación:`);
  for (const d of r.detalles) console.log(`  ${d}`);
}

function imprimirResumen(resultados: Map<string, Resultado>): void {
  const recuperados = [...resultados.values()].filter((r) => r.exito).length;
  console.log(`\n${"=".repeat(60)}`);
  console.log("  RESUMEN — Simulador de Incidentes de Producción");
  console.log("=".repeat(60));
  console.log(`\n  ${recuperados}/${resultados.size} fallos recuperados automáticamente`);
  console.log(`\n  ${"Fallo".padEnd(22)} ${"Estado".padEnd(16)} Estrategia`);
  console.log("  " + "─".repeat(56));
  for (const [tipo, r] of resultados) {
    console.log(`  ${tipo.padEnd(22)} ${(r.exito ? "RECUPERADO" : "FALLIDO").padEnd(16)} ${r.estrategia}`);
  }
  console.log("\n  Lecciones clave:");
  console.log("  • Timeout: retry con jitter previene thundering herd");
  console.log("  • Output malformado: feedback exacto al modelo");
  console.log("  • Context overflow: compresión al 75% de uso, no al 100%");
  console.log("  • Tool fallo: circuit breaker evita cascada de timeouts");
  console.log("  • Budget excedido: degradar modelo antes de abortar");
  console.log("=".repeat(60));
}

// ── main ──────────────────────────────────────────────────────────────────────

function main(): void {
  const args = process.argv.slice(2);
  const fallo = args.includes("--fallo") ? args[args.indexOf("--fallo") + 1] : "todos";

  console.log(`\n${"=".repeat(60)}`);
  console.log("  SIMULADOR DE INCIDENTES DE PRODUCCIÓN");
  console.log(`  Fallo: ${fallo}`);
  console.log("=".repeat(60));

  const todos = ["timeout", "output_malformado", "context_overflow", "tool_fallo", "budget_excedido"];
  const fallos = fallo === "todos" ? todos : [fallo];

  const cb = new CircuitBreaker("search_docs", 3);
  (cb as any).fallosExterno = 3; // pre-carga 3 fallos para el escenario

  const resultados = new Map<string, Resultado>();
  for (const f of fallos) {
    let r: Resultado;
    if (f === "timeout") r = recuperarTimeout();
    else if (f === "output_malformado") r = recuperarOutputMalformado();
    else if (f === "context_overflow") r = recuperarContextOverflow(7200, 8192);
    else if (f === "tool_fallo") r = recuperarToolFallo(cb);
    else if (f === "budget_excedido") r = recuperarBudgetExcedido(0.082, 0.06);
    else continue;
    imprimirResultado(f, r);
    resultados.set(f, r);
  }

  if (resultados.size > 1) imprimirResumen(resultados);
}

main();
